### PR TITLE
Adding OpenMP parallelism to RomB&S

### PR DESCRIPTION
**Pull request by Rbravo555**
This PR acts on two functions of the RomB&S:

  -BuildAndSolve
  -ProjectToFineBasis

Both of these functions could potentially benefit from parallelism, specially the latter, as the skin in an H-ROM model contains still a relatively large number of Conditions.

When approved, the same implementation is to be added to other scripts which also project some elemental contribution to the reduced space.

### CODE REVIEW

File: applications/RomApplication/costum_strategies/rom_builder_and_solver.h [hunk 1]
```diff
@@ -278,9 +278,13 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
...
         TSparseSpace::SetToZero(rX);
-        for (const auto &node : rNodes){
-            unsigned int node_aux_id = node.GetValue(AUX_ID);
-            const auto &nodal_rom_basis = node.GetValue(ROM_BASIS);
+        auto nodes_begin = rNodes.begin();
+        auto nodes_number = rNodes.size();    
+        #pragma omp parallel for firstprivate(nodes_number)
+        for (int kkk = 0; kkk < nodes_number; kkk++ ){
+            auto it = nodes_begin + kkk;
+            unsigned int node_aux_id = it->GetValue(AUX_ID);
```
Comment (philbucher) [inline applications/RomApplication/costum_strategies/rom_builder_and_solver.h:286]: I recommend testing this also in `FullDebug`

File: applications/RomApplication/costum_strategies/rom_builder_and_solver.h [hunk 2]
```diff
@@ -278,9 +278,13 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
         TSystemVectorType &rX)
     {
         TSparseSpace::SetToZero(rX);
-        for (const auto &node : rNodes){
-            unsigned int node_aux_id = node.GetValue(AUX_ID);
-            const auto &nodal_rom_basis = node.GetValue(ROM_BASIS);
+        auto nodes_begin = rNodes.begin();
```
Comment (RiccardoRossi) [inline applications/RomApplication/costum_strategies/rom_builder_and_solver.h]: const

and also should be both firstprivate

File: applications/RomApplication/costum_strategies/rom_builder_and_solver.h [hunk 3]
```diff
@@ -346,80 +352,91 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
...
+                    //assemble the elemental contribution - here is where the ROM acts
+                    //compute the elemental reduction matrix Phi
+                    const auto &geom = it_el->GetGeometry();
+                    Matrix PhiElemental(geom.size() * mNodalDofs, mRomDofs);
+
+                    for (unsigned int i = 0; i < geom.size(); ++i){
+                        const Matrix &rom_nodal_basis = geom[i].GetValue(ROM_BASIS);
+                        for (unsigned int k = 0; k < rom_nodal_basis.size1(); ++k){
+                            if (dofs[i * mNodalDofs + k]->IsFixed())
+                                row(PhiElemental, i * mNodalDofs + k) = ZeroVector(PhiElemental.size2());
```
Comment (RiccardoRossi) [inline applications/RomApplication/costum_strategies/rom_builder_and_solver.h]: noalias(row ....)

Comment (RiccardoRossi) [reply applications/RomApplication/costum_strategies/rom_builder_and_solver.h]: or row(...).clear()

File: applications/RomApplication/costum_strategies/rom_builder_and_solver.h [hunk 4]
```diff
@@ -346,80 +352,91 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
...
+                    //assemble the elemental contribution - here is where the ROM acts
+                    //compute the elemental reduction matrix Phi
+                    const auto &r_geom = it->GetGeometry();
+                    Matrix PhiElemental(r_geom.size() * mNodalDofs, mRomDofs);
+
+                    for (unsigned int i = 0; i < r_geom.size(); ++i){
+                        const Matrix &rom_nodal_basis = r_geom[i].GetValue(ROM_BASIS);
+                        for (unsigned int k = 0; k < rom_nodal_basis.size1(); ++k){
+                            if (dofs[i * mNodalDofs + k]->IsFixed())
+                                row(PhiElemental, i * mNodalDofs + k) = ZeroVector(PhiElemental.size2());
```
Comment (RiccardoRossi) [inline applications/RomApplication/costum_strategies/rom_builder_and_solver.h]: same as above

File: applications/RomApplication/costum_strategies/rom_builder_and_solver.h (hunk 2 above)
Comment (Rbravo555) [reply applications/RomApplication/costum_strategies/rom_builder_and_solver.h]: In line 355 we have a similar case, and have not make const nor firstprivate the ModelPart.ElementsBegin() nor ModelPart.ConditionsBegin() variables

Comment (philbucher) [reply applications/RomApplication/costum_strategies/rom_builder_and_solver.h]: so the `const` methods are missing?

**Comment by Rbravo555** [comment]
since this PR has diverged due to new concepts added. I am erasing this branch and creating another one derived from #6534

**Pull request closed without merging**